cities/queries/causal_insight.py

```diff
@@ -0,0 +1,413 @@
+        self.outcome_dataset = outcome_dataset
+        self.intervention_dataset = intervention_dataset
+        self.root = find_repo_root()
+        self.num_samples = num_samples
+
+        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+    def load_guide(self, forward_shift):
+        pyro.clear_param_store()
+        guide_name = (
+            f"{self.intervention_dataset}_{self.outcome_dataset}_{forward_shift}"
+        )
+        guide_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_guide.pkl"
+        )
+
+        with open(guide_path, "rb") as file:
+            self.guide = dill.load(file)
+        param_path = os.path.join(
+            self.root, "data/model_guides", f"{guide_name}_params.pth"
+        )
+
+        pyro.get_param_store().load(param_path)
+
+        self.forward_shift = forward_shift
+
+    def generate_samples(self):
+        self.data = prep_wide_data_for_inference(
+            outcome_dataset=self.outcome_dataset,
+            intervention_dataset=self.intervention_dataset,
+            forward_shift=self.forward_shift,
+        )
+        self.model = model_cities_interaction
+
+        self.model_args = self.data["model_args"]
+
+        self.model_conditioned = pyro.condition(
+            self.model,
+            data={"T": self.data["t"], "Y": self.data["y"], "X": self.data["x"]},
+        )
+
+        self.predictive = pyro.infer.Predictive(
+            model=self.model,
+            guide=self.guide,
+            num_samples=self.num_samples,
+            parallel=True,
+        )
+        self.samples = self.predictive(*self.model_args)
+
+        # idexing and gathering with mwc in this context
+        # seems to fail, calculating the expected diff made by the intervention manually
+        # wrt to actual observed outcomes rather than predicting outcomes themselves
+        # effectively keeping the noise fixed and focusing on a counterfactual claim
+
+        # TODO possible delete in the current strategy deemed uncontroversial
+        # else:
+        #     if not isinstance(intervened_value, torch.Tensor):
+        #         intervened_value = torch.tensor(intervened_value, device=self.device)
+        #     intervened_expanded = intervened_value.expand_as(self.data['t'])
+
+        #     with MultiWorldCounterfactual(first_available_dim=-6) as mwc:
+        #         with do(actions = dict(T = intervened_expanded)):
+        #                 self.predictive = pyro.infer.Predictive(model=self.model, guide=self.guide,
+        #                         num_samples=self.num_samples, parallel=True)
+        #                 self.samples = self.predictive(*self.model_args)
+        #     self.mwc = mwc
+
+    def generate_tensed_samples(self):
+        self.tensed_samples = {}
+        self.tensed_tau_samples = {}
+
+        for shift in [1, 2, 3]:
+            self.load_guide(shift)
```

Comment:
This is a bit odd to me, because e.g. `self.guide` will just be whatever you happened to load last. See file-level comment for how to restructure this.

Reply:
yeh, but guides are not used after the predictions are generated, so I don't really care which guide is left hanging. Will take a look at the detailed comment though.

Reply:
This one I think would be most robustly handled by explicitly handling the currently implicit 1:M that's going on here.

> get_fips_prediction obviously needs to take its arguments, and I think you're setting them as attributes for the sake of plotting? In this case, I would make another object (or struct type thing) that is constructed in the get_fips_prediction method and is the target of plotting. That would better capture the 1:M relationship between this one class and the many queries it can answer (as opposed to internally hanging on to information needed to reconstruct just the most recent query)`.